isobus/src/can_internal_control_function.cpp

```diff
@@ -11,64 +11,378 @@
+		{
+			case State::None:
+			{
+				set_current_state(State::WaitForClaim);
+			}
+			break;
+
+			case State::WaitForClaim:
+			{
+				if (SystemTiming::time_expired_ms(stateChangeTimestamp_ms, randomClaimDelay_ms))
+				{
+					set_current_state(State::SendRequestForClaim);
+				}
+			}
+			break;
+
+			case State::SendRequestForClaim:
+			{
+				if (send_request_to_claim())
+				{
+					set_current_state(State::WaitForRequestContentionPeriod);
+				}
+			}
+			break;
+
+			case State::WaitForRequestContentionPeriod:
+			{
+				if (SystemTiming::time_expired_ms(stateChangeTimestamp_ms, ADDRESS_CONTENTION_TIME_MS))
+				{
+					std::shared_ptr<ControlFunction> deviceAtOurPreferredAddress;
+					if (NULL_CAN_ADDRESS != preferredAddress)
+					{
+						deviceAtOurPreferredAddress = CANNetworkManager::CANNetwork.get_control_function(get_can_port(), preferredAddress);
+					}
+
+					// Time to find a free address
+					if ((nullptr == deviceAtOurPreferredAddress) && (NULL_CAN_ADDRESS != preferredAddress))
+					{
+						// Our address is free. This is the best outcome. Claim it.
+						set_current_state(State::SendPreferredAddressClaim);
+					}
+					else if (get_NAME().get_arbitrary_address_capable())
+					{
+						// Our address is not free, but we can tolerate an arbitrary address
+						set_current_state(State::SendArbitraryAddressClaim);
+					}
+					else if ((!get_NAME().get_arbitrary_address_capable()) &&
+					         (nullptr != deviceAtOurPreferredAddress) &&
+					         (deviceAtOurPreferredAddress->get_NAME().get_full_name() > get_NAME().get_full_name()))
+					{
+						// Our address is not free, we cannot be at an arbitrary address, but address is contendable
+						set_current_state(State::ContendForPreferredAddress);
+					}
+					else
+					{
+						// Can't claim because we cannot tolerate an arbitrary address, and the CF at that spot wins contention
+						set_current_state(State::UnableToClaim);
+						LOG_ERROR("[AC]: Internal control function %016llx failed to claim its preferred address %u on channel %u, as it cannot tolerate for an arbitrary address and there is already a CF at the preferred address that wins contention.",
+						          get_NAME().get_full_name(),
+						          preferredAddress,
+						          get_can_port());
+					}
+				}
+			}
+			break;
+
+			case State::SendPreferredAddressClaim:
+			{
+				if (send_address_claim(preferredAddress))
+				{
+					LOG_DEBUG("[AC]: Internal control function %016llx has claimed address %u on channel %u",
+					          get_NAME().get_full_name(),
+					          get_preferred_address(),
+					          get_can_port());
+					hasClaimedAddress = true;
+					set_current_state(State::AddressClaimingComplete);
+				}
+				else
+				{
+					set_current_state(State::None);
+				}
+			}
+			break;
+
+			case State::SendArbitraryAddressClaim:
+			{
+				// Search the range of generally available addresses
+				for (std::uint8_t i = 128; i <= 235; i++)
```

Comment:
Technically this range depends on the NAME's industry group. We may want to enhance this to properly adjust the range at some point. For example, the construction industry group only allows for 128 to 207, with 208 and up being "reserved". We can reference the ranges on ISOBUS.net https://www.isobus.net/isobus/sourceAddress

Reply:
Added in https://github.com/Open-Agriculture/AgIsoStack-plus-plus/pull/463/commits/799f30292fae15e469eb85420a8630ce50c567ed